### DISCUSSION TITLE
Use cases for WebHooks in Arena

### DISCUSSION THREAD

**Discussion by delros** [unitedworldwrestling/arena-public#158] [General] [answered]
Hi everyone, 

The capability to use the webhooks is present in Arena and the respective endpoints defined in the specification. Still, I'm struggling to find any related documentation or examples for events that can be subscribed or typical use cases when and where these webhooks might be used. 

![image](https://github.com/unitedworldwrestling/arena-public/assets/1046895/be8a842c-b58e-4daf-b62c-feddb5e5b577)

I would assume that the webhooks are used by "companion" apps and/or for synchronization as per the master-slave synchronization architecture of the system, but in case if we can create a custom listener of the events, that means bespoke external "companion" apps also can be created to extend the native functionality, which looks prominent enough.

Do we have any examples of the webhooks usage or documentation about supported events and hook payloads? Could you please share any related insights regarding these capabilities?

**Comment by eveyrat** [comment, marked as answer]
Hi @delros,

You can configure Webhooks through the UI available at `/bracket/webhook/`. You can "listen" for different events (Insert, Update, Delete), on different entities (Fight, SportEvent, etc.), and get notified on the endpoint of your choice.
<img width="1177" alt="image" src="https://github.com/unitedworldwrestling/arena-public/assets/403226/92619ed9-7fb8-4bce-8575-5032704ff404">


Arena will trigger a POST request to your endpoint.
I'd suggest creating a Test Webhook for inspecting the payload using a service like https://webhook.site

The payload is very minimal. It contains the Entity, ID and action. Then up to you to do further API calls depending on your use case:
```
{
  "action": "<action>",
  "entity": "<entity>",
  "id": "<entity_id>"
}
```

Example:
```
{
  "action": "update",
  "entity": "Fight",
  "id": "1eec01fa-25d8-6c3a-9bcc-5b5367a7e936"
}
```

The built-in master-slave sync is not using Webhooks but the purpose is to give the flexibility to anyone to integrate Arena in their own workflows.
For instance syncing with a third party app, or At UWW, we use it for CDN cache invalidation.

**Reply by delros** [reply]
That's exactly what I was interested in -  `/bracket/webhook/`. I'm not familiar enough at the moment with Arean UI and how it operates (besides recorded videos of the workshops shared in a repo), so this guidance provides the essential direction for future investigation and learning. Thanks a lot @eveyrat

**Reply by delros** [reply]
@eveyrat, could you please advise if there are any limitations for a specific setup required for a webhook pointed to a localhost service? testing the setup on local arena, where hooks to webhooks.site sent as expected, but no requests to local service instance has been recieved (http://localhost:3002/webhooks/arena), but direct POST call to service endpoint accepted without issues

**Reply by eveyrat** [reply]
Hi @delros,
You'll probably want to proxy localhost using a service like [ngrok](https://ngrok.com/).
Once you've installed the ngrok CLI, just run:
`ngrok http <your-app-port>`

**Reply by csabavirag** [reply]
Hello @delros, you can't use localhost in the URL, because Arena sees `localhost` as its own internal network. I assume you are developing your webhook receiver, so the best way is to use your host computer's IP address instead of localhost.
So instead of http://localhost:3002/webhooks/arena, you should use something like http://192.168.1.50/webhooks/arena where 192.168.1.50 is your LAN IP address.
Of course, if your host computer's IP address changes, you must update the webhook on Arena according to the new IP. (If you can manage your LAN's DNS, you might want to create a DNS entry for your host, thus you can even use that in the webhook config. Or assign a static IP to the host, thus it will never change)

However using ngrok is also an option, this way you can test HTTPS easily.

**Reply by delros** [reply]
@csabavirag, oh, that might work, thanks for the idea, since currently I'm using ngrok for tunneling the hooks to local service (@eveyrat we are on the same page 👍). however, this is not a blocking case, so was wondering if I misunderstood or just miss something. Thanks you both

**Comment by csabavirag** [comment]
Perhaps the easiest way to get in touch with Webhooks is to create one and point to https://webhook.site/ When you visit this page, you will get a dedicated webhook where you can forward actions from Arena. Then you will see the payload being sent to the webhook listener through their web interface.

Use case can be for example:
You have a service/application querying Arena API to display/aggregate data on the fight details. To call Arena for fetching all the fights of a competition for example is resource heavy with every single request your service receives, thus you might want to "cache" the data coming from Arena to avoid unnecessary calls to the backend. Especially when the data is not changed (eg. you are in between 2 sessions, thus no fights are updated).

So to avoid this extra load (and extra delay to every single request), you can have 2 options.
1. Invalidate your cache with predefined timing. Eg. you refresh the fights in every 10, or 30 minutes. Between these refreshes, you render the front end with the cached data. That might causes delays in the displayed information. What about if your next session starts right after you refreshed your fights? In the next 10/30 minutes you will display the "previous" state until the next refresh happens. Acceptable? Take this
2. Setup a Webhook listener in your service and refresh the cache (of course if you structure your cache, you just have to refresh a part of it) whenever you receive a webhook trigger. What kind of webhook trigger you must configure? Enable the Fight insert+update+delete webhook. This way, your service will receive a message with the the ID of the fight (also the message will contain the type of the event - insert/update/delete), so you will be able to refresh this fight only from Arena. Next data rendering will contain the updated info about that fight.

Does it make sense?

**Reply by delros** [reply]
Yup, it makes perfect sense, thanks for the notes and highlights
